lib/AstOfLlbc.ml

```diff
@@ -628,39 +896,77 @@ let fn_ptr_is_opaque env (fn_ptr: C.fn_ptr) =
   | _ ->
       false
 
-let expression_of_fn_ptr env (fn_ptr: C.fn_ptr) =
+(* This is a very core piece of logic that transforms a Rust fn_ptr into a krml AST node that
+   contains type application, const generic applications, and application of trait methods to
+   implement the dictionary-passing style. *)
+let rec expression_of_fn_ptr env depth (fn_ptr: C.fn_ptr) =
   let {
     C.generics = { types = type_args; const_generics = const_generic_args; trait_refs; _ };
     func;
     _
   } = fn_ptr in
 
   (* General case for function calls and trait method calls. *)
-  L.log "Calls" "Visiting call: %s" (Charon.PrintExpressions.fn_ptr_to_string env.format_env fn_ptr);
-  L.log "Calls" "is_array_map: %b" (RustNames.is_array_map env fn_ptr);
-  L.log "Calls" "--> %d type_args, %d const_generics, %d trait_refs"
+  L.log "Calls" "%sVisiting call: %s" depth (Charon.PrintExpressions.fn_ptr_to_string env.format_env fn_ptr);
+  L.log "Calls" "%sis_array_map: %b" depth (RustNames.is_array_map env fn_ptr);
+  L.log "Calls" "%s--> %d type_args, %d const_generics, %d trait_refs" depth
     (List.length type_args) (List.length const_generic_args) (List.length trait_refs);
 
   let type_args, const_generic_args, trait_refs =
     match func with
-    | TraitMethod ({ generics = { types; const_generics; trait_refs; _ }; _ }, _, _) ->
-        types @ type_args, const_generics @ const_generic_args, trait_refs @ trait_refs
+    | TraitMethod ({ generics = { types; const_generics; trait_refs = trait_refs'; _ }; _ }, _, _) ->
+        (* TODO: why is the order fumbled here? *)
```

Comment:
What do you mean?

Reply:
What I meant is that trait_refs' was concatenated in the opposite order as the concatenation of `const_generics` -- this has since then been fixed.